src/agent/pythd/api.rs

```diff
@@ -304,65 +304,127 @@ pub mod rpc {
+
+                    // Perform requests in sequence and gather responses
+                    for request in requests {
+                        let response = self.dispatch_and_catch_error(&request).await;
+                        responses.push(response)
+                    }
+
+                    // Send an array if we're handling a batch
+                    // request, single response object otherwise
+                    if is_batch {
+                        self.send_text(&serde_json::to_string(&responses)?).await?;
+                    } else {
+                        self.send_text(&serde_json::to_string(&responses[0])?)
+                            .await?;
+                    }
+                }
+                // The top-level parsing errors are fine to share with client
+                Err(e) => {
+                    self.send_error(e, None).await?;
+                }
             }
-        }
 
-        async fn dispatch_and_catch_error(
-            &mut self,
-            request: &Request<Method, Value>,
-        ) -> Result<()> {
-            if let Err(err) = self.dispatch(request).await {
-                error!(self.logger, "{:#}", err; "error" => format!("{:?}", err));
-                self.send_error_response(err, request).await?;
-            };
             Ok(())
         }
 
-        async fn parse(
-            &mut self,
-            msg: Message,
-        ) -> core::result::Result<Request<Method, Value>, jrpc::Error<Value>> {
-            let s = msg.to_str().map_err(|()| {
-                // This should never happen, but fail gracefully just in case
-                jrpc::Error::new(
-                    Id::from(0),
-                    jrpc::ErrorCode::ParseError,
-                    "non-text websocket message received",
-                    None,
-                )
-            })?;
-            parse_request::<Method>(s)
+        /// Parse a JSONRPC request object or a batch of them. The
+        /// bool in result informs request handling whether it needs
+        /// to respond with a single object or an array, to prevent
+        /// sending unexpected
+        /// `[{<just one response, but request was not array>}]`
+        /// array payloads.
+        async fn parse(&mut self, msg: Message) -> Result<(Vec<Request<Method, Value>>, bool)> {
+            let s = msg
+                .to_str()
+                .map_err(|_| anyhow!("Could not parse message as text"))?;
+
+            let json_value: Value = serde_json::from_str(s)?;
+            if let Some(array) = json_value.as_array() {
+                // Interpret request as JSON-RPC 2.0 batch if value is an array
+                let mut requests = Vec::with_capacity(array.len());
+                for maybe_request in array {
+                    // parse_request() is the only jrpc parsing method which takes &str
+                    let maybe_request_string = serde_json::to_string(maybe_request)?;
```

Comment:
this comment is confusing. I think you're explaining why you have to re-stringify the json object (there's only one method for parsing requests), but the text isn't explicit enough. I would lead with something like "It's unfortunate that we have to stringify the json object here, but ... "

Reply:
Changed to something along what you propose.